src/vm_pmu.rs

```diff
@@ -368,4 +375,195 @@ impl VmPmuState {
             }
         }
     }
+
+    /// Helper function to call either SBI, or the internal implementation
+    /// for configure_matching_counters() depending on whether Sscdeleg is implemented
+    pub fn configure_matching_counters(
+        &self,
+        counter_index: u64,
+        counter_mask: u64,
+        config_flags: PmuCounterConfigFlags,
+        event_type: PmuEventType,
+        event_data: u64,
+    ) -> SbiResult<u64> {
+        if !CpuInfo::get().has_sscdeleg() {
+            sbi::api::pmu::configure_matching_counters(
+                counter_index,
+                counter_mask,
+                config_flags,
+                event_type,
+                event_data,
+            )
+        } else {
+            fn configure_counter_csrs(
+                config_flags: PmuCounterConfigFlags,
+                event_type: PmuEventType,
+                event_data: u64,
+                counter_index: u64,
+            ) {
+                let mut shpmevent = LocalRegisterCopy::<u64, shpmevent::Register>::new(0);
+                // Ignore vuinh / vsinh from the VM and translate uinh
+                if config_flags.is_uinh() {
+                    shpmevent.modify(shpmevent::vuinh.val(1));
+                }
+                if config_flags.is_sinh() {
+                    shpmevent.modify(shpmevent::sinh.val(1));
+                }
+                match event_type {
+                    PmuEventType::Hardware(_) | PmuEventType::Cache(_) => {
+                        shpmevent.modify(shpmevent::event_selector.val(event_type.raw()))
+                    }
+                    PmuEventType::RawEvent => {
+                        shpmevent.modify(shpmevent::event_selector.val(event_data))
+                    }
+                    _ => unreachable!(),
+                }
+                CSR.scounterselect.set(counter_index);
+                CSR.shpmevent.set(shpmevent.get());
+                if config_flags.is_clear_value() {
+                    CSR.shpmcounter.set(0);
+                }
+                if config_flags.is_auto_start() {
+                    CSR.scountinhibit.read_and_clear_bits(1 << counter_index);
+                }
+            }
+
+            let pmu_info = pmu::PmuInfo::get()?;
+            let counter_mask = pmu_info.filter_counter_mask(counter_index, counter_mask)?;
+            let index = if !config_flags.is_skip_match() {
+                match event_type {
```

Comment:
I think it might be more readable to write this logic like this:

```
let index = if !config_flags.is_skip_match() {
    let iter = CounterMaskIter::new(...);
    match event_type {
        PmuHardware::CpuCycles => {
            if iter.find(|c| c == CYCLES).is_some() && matches!(self.counter_state[CYCLES], NotConfigured) {
                CYCLES
            } else {
                return Err(...);
            }
        }
        PmuHardware::Instructions => { ... /* Same as above but s/CYCLES/INSTRET/ */ }
        _ => iter.find(...).ok_or(...),
    }
} else {
    ...
}?;
...
```